File: decode_zhanshi_server/decode.py
#!/usr/bin/env python3
"""
传世展示页 (zhanshi_server_64.php) 解码工具

用法:
    python3 decode.py <密钥> [加密方式]

例:
    python3 decode.py "your_secret_key"
    python3 decode.py "your_secret_key" aes-256-cbc

需要安装: pip install pycryptodome requests
"""

import sys
import base64
import hashlib
import os

try:
    from Crypto.Cipher import AES, Blowfish, DES3, ARC4
    from Crypto.Util.Padding import unpad
except ImportError:
    print("请先安装 pycryptodome: pip install pycryptodome", file=sys.stderr)
    sys.exit(1)

URL = "https://www.3tyx.top/2/zhanshi_server_64.php"
LOCAL_CACHE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "encrypted_data.txt")


def fetch_data():
    try:
        import requests
        resp = requests.get(URL, timeout=10)
        resp.raise_for_status()
        return resp.text.strip()
    except Exception:
        pass
    try:
        import urllib.request
        resp = urllib.request.urlopen(URL, timeout=10)
        return resp.read().decode().strip()
    except Exception:
        pass
    if os.path.exists(LOCAL_CACHE):
        print("使用本地缓存文件", file=sys.stderr)
        return open(LOCAL_CACHE).read().strip()
    return None


def try_decrypt_aes(decoded, key_bytes, mode, iv=None):
    try:
        if mode == "cbc":
            cipher = AES.new(key_bytes, AES.MODE_CBC, iv)
        elif mode == "ecb":
            cipher = AES.new(key_bytes, AES.MODE_ECB)
        else:
            return None
        result = cipher.decrypt(decoded)
        try:
            result = unpad(result, AES.block_size)
        except ValueError:
            pass
        return result
    except Exception:
        return None


def is_readable(data, threshold=0.5):
    if not data or len(data) < 10:
        return False
    sample = data[:min(200, len(data))]
    printable = sum(1 for b in sample if 32 <= b < 127 or b in (10, 13, 9))
    return (printable / len(sample)) > threshold


def main():
    if len(sys.argv) < 2:
        print("用法: python3 decode.py <密钥> [加密方式]", file=sys.stderr)
        print("示例: python3 decode.py \"my_secret_key\"", file=sys.stderr)
        sys.exit(1)

    key = sys.argv[1]
    specified_method = sys.argv[2].lower() if len(sys.argv) > 2 else None

    print("正在获取加密数据...", file=sys.stderr)
    raw = fetch_data()
    if not raw:
        print("错误: 无法获取数据", file=sys.stderr)
        sys.exit(1)

    decoded = base64.b64decode(raw)
    print(f"Base64 解码后数据长度: {len(decoded)} 字节", file=sys.stderr)

    key_variants = {
        "raw": key.encode("utf-8"),
        "md5_raw": hashlib.md5(key.encode()).digest(),
        "md5_hex": hashlib.md5(key.encode()).hexdigest().encode(),
        "sha256_raw_16": hashlib.sha256(key.encode()).digest()[:16],
        "sha256_raw_32": hashlib.sha256(key.encode()).digest(),
        "sha256_hex": hashlib.sha256(key.encode()).hexdigest().encode(),
    }

    # Also try if key itself is hex
    try:
        key_variants["hex_decoded"] = bytes.fromhex(key)
    except ValueError:
        pass

    methods = (
        [specified_method] if specified_method
        else ["aes-128-cbc", "aes-256-cbc", "aes-128-ecb", "aes-256-ecb", "rc4", "bf-cbc"]
    )

    for method in methods:
        for var_name, key_bytes in key_variants.items():
            result = None

            if method == "aes-128-cbc" and len(key_bytes) >= 16:
                k = key_bytes[:16]
                if len(decoded) > 16 and len(decoded[16:]) % 16 == 0:
                    iv = decoded[:16]
                    result = try_decrypt_aes(decoded[16:], k, "cbc", iv)
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}, IV=数据前16字节", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return
                # zero IV
                if len(decoded) % 16 == 0:
                    result = try_decrypt_aes(decoded, k, "cbc", b'\x00' * 16)
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}, IV=全零", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return

            elif method == "aes-256-cbc" and len(key_bytes) >= 32:
                k = key_bytes[:32]
                if len(decoded) > 16 and len(decoded[16:]) % 16 == 0:
                    iv = decoded[:16]
                    result = try_decrypt_aes(decoded[16:], k, "cbc", iv)
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}, IV=数据前16字节", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return
                if len(decoded) % 16 == 0:
                    result = try_decrypt_aes(decoded, k, "cbc", b'\x00' * 16)
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}, IV=全零", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return

            elif method == "aes-128-ecb" and len(key_bytes) >= 16:
                k = key_bytes[:16]
                if len(decoded) % 16 == 0:
                    result = try_decrypt_aes(decoded, k, "ecb")
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return

            elif method == "aes-256-ecb" and len(key_bytes) >= 32:
                k = key_bytes[:32]
                if len(decoded) % 16 == 0:
                    result = try_decrypt_aes(decoded, k, "ecb")
                    if result and is_readable(result):
                        print(f"解密成功! 方式={method}, 密钥变体={var_name}", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return

            elif method == "rc4":
                try:
                    cipher = ARC4.new(key_bytes)
                    result = cipher.decrypt(decoded)
                    if is_readable(result):
                        print(f"解密成功! 方式=RC4, 密钥变体={var_name}", file=sys.stderr)
                        sys.stdout.buffer.write(result)
                        return
                except Exception:
                    pass

            elif method == "bf-cbc" and 4 <= len(key_bytes) <= 56:
                if len(decoded) > 8 and len(decoded[8:]) % 8 == 0:
                    try:
                        iv = decoded[:8]
                        cipher = Blowfish.new(key_bytes, Blowfish.MODE_CBC, iv)
                        result = cipher.decrypt(decoded[8:])
                        if is_readable(result):
                            print(f"解密成功! 方式=Blowfish-CBC, 密钥变体={var_name}", file=sys.stderr)
                            sys.stdout.buffer.write(result)
                            return
                    except Exception:
                        pass

    # Last resort: try decompression
    import zlib, gzip
    for name, fn in [("gzinflate", lambda d: zlib.decompress(d, -15)),
                     ("gzuncompress", zlib.decompress),
                     ("gzdecode", gzip.decompress)]:
        try:
            result = fn(decoded)
            print(f"{name} 解压成功!", file=sys.stderr)
            sys.stdout.buffer.write(result)
            return
        except Exception:
            pass

    print("\n所有方式均未成功解密。", file=sys.stderr)
    print("请检查密钥是否正确，或指定正确的加密方式。", file=sys.stderr)
    print(f"数据特征:", file=sys.stderr)
    print(f"  - 长度: {len(decoded)} 字节 (16字节对齐)", file=sys.stderr)
    print(f"  - 前8字节 ASCII: {decoded[:8]}", file=sys.stderr)
    print(f"  - 前16字节 HEX: {decoded[:16].hex()}", file=sys.stderr)
    sys.exit(1)


if __name__ == "__main__":
    main()

File: decode_zhanshi_server/decode.php
<?php
/**
 * 传世展示页 (zhanshi_server_64.php) 解码工具
 *
 * 用法:
 *   php decode.php <密钥> [加密方式]
 *
 * 例:
 *   php decode.php "your_secret_key"
 *   php decode.php "your_secret_key" aes-256-cbc
 *   php decode.php "your_secret_key" aes-128-cbc
 *
 * 支持的加密方式 (默认尝试全部):
 *   aes-128-cbc, aes-256-cbc, aes-128-ecb, aes-256-ecb
 */

$url = 'https://www.3tyx.top/2/zhanshi_server_64.php';

if ($argc < 2) {
    fwrite(STDERR, "用法: php decode.php <密钥> [加密方式]\n");
    fwrite(STDERR, "示例: php decode.php \"my_secret_key\"\n");
    fwrite(STDERR, "      php decode.php \"my_secret_key\" aes-256-cbc\n\n");
    fwrite(STDERR, "支持的加密方式: aes-128-cbc, aes-256-cbc, aes-128-ecb, aes-256-ecb\n");
    fwrite(STDERR, "不指定加密方式则自动尝试所有常见方式。\n");
    exit(1);
}

$key = $argv[1];
$specifiedMethod = isset($argv[2]) ? strtolower(trim($argv[2])) : null;

fwrite(STDERR, "正在获取加密数据...\n");

$raw = @file_get_contents($url);
if ($raw === false) {
    $localFile = __DIR__ . '/encrypted_data.txt';
    if (file_exists($localFile)) {
        $raw = file_get_contents($localFile);
        fwrite(STDERR, "使用本地缓存文件 encrypted_data.txt\n");
    } else {
        fwrite(STDERR, "错误: 无法获取数据\n");
        exit(1);
    }
}

$raw = trim($raw);
$decoded = base64_decode($raw, true);
if ($decoded === false) {
    fwrite(STDERR, "错误: Base64 解码失败\n");
    exit(1);
}

fwrite(STDERR, "Base64 解码后数据长度: " . strlen($decoded) . " 字节\n");

$methods = $specifiedMethod
    ? [$specifiedMethod]
    : ['aes-128-cbc', 'aes-256-cbc', 'aes-128-ecb', 'aes-256-ecb',
       'aes-128-cfb', 'aes-256-cfb', 'aes-128-ofb', 'aes-256-ofb',
       'bf-cbc', 'bf-ecb', 'des-ede3-cbc', 'des-ede3-ecb', 'rc4'];

$keyVariants = [
    'raw'          => $key,
    'md5_hex'      => md5($key),
    'md5_raw'      => md5($key, true),
    'sha256_hex'   => hash('sha256', $key),
    'sha256_raw'   => hash('sha256', $key, true),
];

$success = false;

foreach ($methods as $method) {
    if (!in_array($method, openssl_get_cipher_methods(true))) {
        continue;
    }

    $ivLen = openssl_cipher_iv_length($method);

    foreach ($keyVariants as $varName => $keyVal) {

        // Strategy A: IV prepended to ciphertext
        if ($ivLen > 0 && strlen($decoded) > $ivLen) {
            $iv = substr($decoded, 0, $ivLen);
            $ct = substr($decoded, $ivLen);

            $result = @openssl_decrypt($ct, $method, $keyVal, OPENSSL_RAW_DATA, $iv);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName, IV=数据前{$ivLen}字节\n");
                echo $result;
                $success = true;
                break 2;
            }

            // Try with base64 input (openssl default)
            $result = @openssl_decrypt(base64_encode($ct), $method, $keyVal, 0, $iv);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName (base64模式)\n");
                echo $result;
                $success = true;
                break 2;
            }
        }

        // Strategy B: zero IV
        if ($ivLen > 0) {
            $iv = str_repeat("\0", $ivLen);
            $result = @openssl_decrypt($decoded, $method, $keyVal, OPENSSL_RAW_DATA, $iv);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName, IV=全零\n");
                echo $result;
                $success = true;
                break 2;
            }

            $result = @openssl_decrypt($raw, $method, $keyVal, 0, $iv);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName (base64输入, 零IV)\n");
                echo $result;
                $success = true;
                break 2;
            }
        }

        // Strategy C: ECB / stream ciphers (no IV)
        if ($ivLen === 0) {
            $result = @openssl_decrypt($decoded, $method, $keyVal, OPENSSL_RAW_DATA);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName (无IV)\n");
                echo $result;
                $success = true;
                break 2;
            }

            $result = @openssl_decrypt($raw, $method, $keyVal, 0);
            if ($result !== false && strlen($result) > 10) {
                fwrite(STDERR, "解密成功! 方式=$method, 密钥变体=$varName (base64输入)\n");
                echo $result;
                $success = true;
                break 2;
            }
        }
    }
}

if (!$success) {
    // Try gzinflate / gzuncompress on the decoded data
    $gz = @gzinflate($decoded);
    if ($gz !== false) {
        fwrite(STDERR, "gzinflate 解压成功!\n");
        echo $gz;
        exit(0);
    }

    $gz = @gzuncompress($decoded);
    if ($gz !== false) {
        fwrite(STDERR, "gzuncompress 解压成功!\n");
        echo $gz;
        exit(0);
    }

    $gz = @gzdecode($decoded);
    if ($gz !== false) {
        fwrite(STDERR, "gzdecode 解压成功!\n");
        echo $gz;
        exit(0);
    }

    fwrite(STDERR, "\n所有方式均未成功解密。\n");
    fwrite(STDERR, "请检查密钥是否正确，或指定正确的加密方式。\n");
    fwrite(STDERR, "数据特征:\n");
    fwrite(STDERR, "  - 长度: " . strlen($decoded) . " 字节 (16字节对齐)\n");
    fwrite(STDERR, "  - 前8字节 ASCII: " . substr($decoded, 0, 8) . "\n");
    fwrite(STDERR, "  - 前16字节 HEX: " . bin2hex(substr($decoded, 0, 16)) . "\n");
    exit(1);
}
